src/operator/sum.rs

```diff
@@ -0,0 +1,215 @@
+//! N-ary plus operator.
+
+use crate::{
+    algebra::{AddAssignByRef, AddByRef, HasZero, WithNumEntries},
+    circuit::{
+        operator_traits::{NaryOperator, Operator},
+        Circuit, OwnershipPreference, Scope, Stream,
+    },
+};
+use std::{
+    borrow::Cow,
+    cmp::Reverse,
+    iter::once,
+    mem::{take, ManuallyDrop},
+    ops::Add,
+};
+
+impl<P, D> Stream<Circuit<P>, D>
+where
+    P: Clone + 'static,
+    D: Add<Output = D> + AddByRef + AddAssignByRef + Clone + HasZero + WithNumEntries + 'static,
+{
+    /// Apply the [`Sum`] operator to `self` and all streams in `streams`.
+    pub fn sum<'a, I>(&'a self, streams: I) -> Stream<Circuit<P>, D>
+    where
+        I: IntoIterator<Item = &'a Self>,
+    {
+        self.circuit()
+            .add_nary_operator(Sum::new(), once(self).chain(streams))
+    }
+}
+
+/// Operator that computes the sum of values across all its input streams at
+/// each timestamp.
+pub struct Sum<D>
+where
+    D: Clone + 'static,
+{
+    // Vector of input values.  Keep it here to reuse the allocation across
+    // operator invocations.
+    inputs: Vec<Cow<'static, D>>,
+}
+
+impl<D> Sum<D>
+where
+    D: Clone + 'static,
+{
+    pub fn new() -> Self {
+        Self { inputs: Vec::new() }
+    }
+}
+
+impl<D> Default for Sum<D>
+where
+    D: Clone + 'static,
+{
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<D> Operator for Sum<D>
+where
+    D: Clone + 'static,
+{
+    fn name(&self) -> Cow<'static, str> {
+        Cow::from("Sum")
+    }
+
+    fn clock_start(&mut self, _scope: Scope) {}
+    fn clock_end(&mut self, _scope: Scope) {}
+}
+
+impl<D> NaryOperator<D, D> for Sum<D>
+where
+    D: Add<Output = D> + AddAssignByRef + Clone + HasZero + WithNumEntries + 'static,
+{
+    fn eval<'a, Iter>(&mut self, inputs: Iter) -> D
+    where
+        Iter: Iterator<Item = Cow<'a, D>>,
+    {
+        let mut input_vec: Vec<Cow<'a, _>> = unsafe {
+            let mut buffer = ManuallyDrop::new(take(&mut self.inputs));
+            let (ptr, len, cap) = (buffer.as_mut_ptr(), buffer.len(), buffer.capacity());
+            Vec::from_raw_parts(ptr, len, cap)
+        };
+
+        assert!(input_vec.is_empty());
+
+        for input in inputs {
+            input_vec.push(input);
+        }
+
+        input_vec.sort_by_key(|x| Reverse(x.num_entries()));
+
+        let mut res = D::zero();
+        for input in input_vec.drain(..) {
+            if res.is_zero() {
+                res = match input {
+                    Cow::Borrowed(v) => v.clone(),
+                    Cow::Owned(v) => v,
+                };
+            } else {
+                match input {
+                    Cow::Borrowed(v) => res.add_assign_by_ref(v),
+                    Cow::Owned(v) => {
+                        res = res + v;
+                    }
```

Comment:
We might as well require `AddAssign` instead of `Add`
```suggestion
                    Cow::Owned(v) => res = res + v,
```

Reply:
How's `AddAssign` better?  It's seems less flexible performance-wise, as it doesn't let you pick the side to add to.

Reply:
That's fair I guess